pkg/build/controller/controller.go

```diff
@@ -108,6 +163,44 @@ func (bc *BuildController) HandleBuild(build *buildapi.Build) error {
 	return nil
 }
 
+func (bc *BuildController) nextBuild(namespace, buildConfigName string) (*buildapi.Build, error) {
+	hasRunningBuilds := false
+	builds, err := buildutil.BuildConfigBuilds(bc.OSClient.Builds(namespace), buildConfigName, func(b buildapi.Build) bool {
+		// Detect if there are running builds
+		switch b.Status.Phase {
+		case buildapi.BuildPhaseRunning, buildapi.BuildPhasePending:
+			hasRunningBuilds = true
+			return false
+		}
+		// Only 'new' build can be scheduled to run next
+		return b.Status.Phase == buildapi.BuildPhaseNew
+	})
+	if hasRunningBuilds || err != nil {
+		return nil, err
+	}
+	var (
+		nextBuild           *buildapi.Build
+		previousBuildNumber int64
+	)
+	for i, b := range builds.Items {
+		buildNumber, err := buildutil.BuildNumber(&b)
+		if err != nil {
+			glog.Errorf("Error getting build number: %v", err)
+			continue
+		}
+		if previousBuildNumber == 0 {
+			nextBuild = &builds.Items[i]
+			previousBuildNumber = buildNumber
+			continue
+		}
```

Comment:
I'll comment here (without digging further) that Jordan's scenario is
probably the one I would expect to see most often in practice on
OpenShift.  I don't think most people care if 3 triggers fire and result in
the last build.

On Sun, Apr 17, 2016 at 4:12 PM, Michal Fojtik <notifications@github.com>
wrote:

> In pkg/build/controller/controller.go
> <https://github.com/openshift/origin/pull/8453#discussion_r59991583>:
>
> > +	}
> > +	var (
> > +		nextBuild           *buildapi.Build
> > +		previousBuildNumber int64
> > +	)
> > +	for i, b := range builds.Items {
> > +		buildNumber, err := buildutil.BuildNumber(&b)
> > +		if err != nil {
> > +			glog.Errorf("Error getting build number: %v", err)
> > +			continue
> > +		}
> > +		if previousBuildNumber == 0 {
> > +			nextBuild = &builds.Items[i]
> > +			previousBuildNumber = buildNumber
> > +			continue
> > +		}
>
> yes, we start with 1, I will rework this :-)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/8453/files/68d0bdac5916b17673695e06451439bfa55941b9#r59991583>
>